modules/vmseries/main.tf

```diff
@@ -1,128 +1,122 @@
+      username   = var.authentication.username
+      public_key = admin_ssh_key.value
     }
   }
 
-  storage_os_disk {
-    create_option     = "FromImage"
-    name              = coalesce(var.os_disk_name, "${var.name}-disk")
-    managed_disk_type = var.managed_disk_type
-    os_type           = "Linux"
-    caching           = "ReadWrite"
+  os_disk {
+    name                   = var.virtual_machine.disk_name
+    storage_account_type   = var.virtual_machine.disk_type
+    caching                = "ReadWrite"
+    disk_encryption_set_id = var.virtual_machine.disk_encryption_set_id
   }
 
-  delete_os_disk_on_termination    = true
-  delete_data_disks_on_termination = true
+  source_image_id = var.image.custom_id
 
-  os_profile {
-    computer_name  = var.name
-    admin_username = var.username
-    admin_password = var.password
-    custom_data    = var.bootstrap_options
+  dynamic "source_image_reference" {
+    for_each = var.image.custom_id == null ? [1] : []
+    content {
+      publisher = var.image.publisher
+      offer     = var.image.offer
+      sku       = var.image.sku
+      version   = var.image.version
+    }
   }
 
-  os_profile_linux_config {
-    disable_password_authentication = var.password == null ? true : false
-    dynamic "ssh_keys" {
-      for_each = var.ssh_keys
-      content {
-        key_data = ssh_keys.value
-        path     = "/home/${var.username}/.ssh/authorized_keys"
-      }
+  dynamic "plan" {
+    for_each = var.image.enable_marketplace_plan ? [1] : []
+
+    content {
+      name      = var.image.sku
+      publisher = var.image.publisher
+      product   = var.image.offer
     }
   }
 
+  custom_data = var.virtual_machine.bootstrap_options == null ? null : base64encode(var.virtual_machine.bootstrap_options)
+
+
   # After converting to azurerm_linux_virtual_machine, an empty block boot_diagnostics {} will use managed storage. Want.
   # 2.36 in required_providers per https://github.com/terraform-providers/terraform-provider-azurerm/pull/8917
-  dynamic "boot_diagnostics" {
-    for_each = var.diagnostics_storage_uri != null ? ["one"] : []
-
-    content {
-      enabled     = true
-      storage_uri = var.diagnostics_storage_uri
-    }
+  boot_diagnostics {
+    storage_account_uri = var.virtual_machine.diagnostics_storage_uri
   }
 
   identity {
-    type         = var.identity_type
-    identity_ids = var.identity_ids
+    type         = var.virtual_machine.identity_type
+    identity_ids = var.virtual_machine.identity_ids
   }
 }
+
+resource "azurerm_network_interface_backend_address_pool_association" "this" {
+  for_each = { for v in var.interfaces : v.name => v.lb_backend_pool_id if v.lb_backend_pool_id != null }
+
+  backend_address_pool_id = each.value
+  ip_configuration_name   = azurerm_network_interface.this[each.key].ip_configuration[0].name
+  network_interface_id    = azurerm_network_interface.this[each.key].id
+
+  depends_on = [
+    azurerm_network_interface.this,
```

Comment:
Do we need directly pass dependency `azurerm_network_interface.this` if we have ` ip_configuration_name   = azurerm_network_interface.this[each.key].ip_configuration[0].name` ?

Reply:
yes, this was causing problem with terraform destroy see [here](https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/commit/d7f7116bc027b1608e0032c88cd6ee0bb125e9ba)